### PR TITLE
added explicit forall notation to make TypeApplications easier to use

### PR DESCRIPTION
**Pull request by Vlix**
This is a PR for issue #869 

---

Before submitting your PR, check that you've:

- [x] Bumped the version number

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

### CODE REVIEW

**Comment by Vlix** [comment]
It's failing because of something in MongoDB?

**Comment by parsonsmatt** [comment]
That's a really strange failure...

**Comment by Vlix** [comment]
All builds using stack works no problem, but built with cabal, it breaks on a missing `MonadFail` constraint.

**Comment by parsonsmatt** [comment]
I suspect we're missing an upper bound somewhere...

**Comment by parsonsmatt** [comment]
Try merging master - there's an upper bound in place to prevent the cabal solver from picking that version.

File: persistent/ChangeLog.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog for persistent
 
+## 2.10.5
```
Comment (parsonsmatt) [inline persistent/ChangeLog.md]: This is a breaking change and requires a major version bump
```suggestion
## 2.11.0
```

File: persistent/persistent.cabal [hunk 2]
```diff
@@ -1,5 +1,5 @@
 name:            persistent
-version:         2.10.4
+version:         2.10.5
```
Comment (parsonsmatt) [inline persistent/persistent.cabal]:
```suggestion
version:         2.11.0
```

**Comment by Vlix** [comment]
I'm trying to add some tests, but it's all really annoying.
Also, all of the methods from typeclasses have the `backend` type variable as number one, because that's what's in the typeclass definition. Don't really know if there'd be a way to make that more user friendly...

**Comment by Vlix** [comment]
I guess one way to do it, would be to not export the class methods in `Database.Persist`, and make new functions that have the same name so anyone importing `Database.Persist` will be able to use the @-notation and anyone making their own (e.g. `PersistStoreWrite`) instance can use `Database.Persist.Class`

**Comment by parsonsmatt** [comment]
I wouldn't worry about tests for this. It's not like it can be broken.

You'd only really want to specify the `rec` where it could be ambiguous, like `selectList` and `deleteWhere`. I wouldn't worry too much about duplicating the names. eg `insert` can't be used polymorphically so it wouldn't make sense to specify the type.

**Comment by Vlix** [comment]
Well, I did anyway.

But yeah, `selectList` will have to still be used as `selectList @_ @Table`, since the `backend` parameter is first, because of the fact that it's a typeclass method. Same with `deleteWhere @_ @Table`. :/

**Comment by parsonsmatt** [comment]
I think `selectList` is fine since it's not a class method. `selectFirst`, `count`, `selectSource` are all affected though :\ `PersistQueryWrite` is also affected, annoying.

Well.

Hmm! Maybe it is worthwhile to have the class methods separate from the functions in the API. It's a breaking change, but *all* of this is a breaking change anyway.

**Comment by Vlix** [comment]
Ah true, `selectList` was outside of the typeclass.

Do you want me to add the `select*` and `delete*` functions in `Database.Persist`?

**Comment by parsonsmatt** [comment]
looking forward to merging this <3

**Pull request merged**